src/transformStateWithClones.js

```diff
@@ -6,8 +6,29 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      action.keysToRemove.forEach((key) => {
+        delete currentState[key];
+      });
+    }
+
+    // Push a clone of the current state to the history
+    stateHistory.push({ ...currentState });
```

Comment:
Instead of using an if-else chain, consider using a switch statement to handle the different action types as it is more suitable for this scenario with limited conditions. This will improve the readability and maintainability of your code.